### PR TITLE
Add jsonschemas

### PR DESCRIPTION
**Pull request by atlaskerr**
This commit adds jsonschemas defined in the distribution spec.

Test fixture generation, file embeding, and schema validation have been
added to the Makefile as well.

Signed-off-by: Atlas Kerr <atlaskerr@gmail.com>

### CODE REVIEW

**Comment by vbatts** [comment]
Big effort here on getting this prepped. The biggest thing I worry about here, because it bit us in the image-spec, is ensuring that every tweak in the spec language was lock-step reflected in the jsonschema, and that unit/integration testing ensuring there were no regressions along the way.

The other (more unfortunate) factor that played in was: many of the data structures are directly related to Golang structures and marshalling.

Every language handles marshelled structs differently and inevitably there is something that is cleaner in one language and inversely more difficult to handle in another. 

Thoughts there?

File: schema/errors.libsonnet [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  errors: 'This property contains a list of errors messages returned from the server due to an invalid request.',
+};
+
+// Error Schema
+local errors(output=JSV7) = {
+
+  local code = {
+    type: 'string',
+    enum: [
+      'BLOB_UNKNOWN',
```
Comment (stevvooe) [inline schema/errors.libsonnet]: This is why I don't like this stuff. What if we add a new error code? This validator fails if it encounters that. This is not an enum necessarily. There are set of known values and unknown future values.

File: schema/errors.libsonnet [hunk 2]
```diff
@@ -0,0 +1,76 @@
+//   Copyright 2018 The Linux Foundation
```
Comment (stevvooe) [inline schema/errors.libsonnet]: Shouldn't this be the OCI Authors?

File: schema/errors.libsonnet [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      'NAME_UNKNOWN',
+      'SIZE_INVALID',
+      'TAG_INVALID',
+      'UNAUTHORIZED',
+      'DENIED',
+      'UNSUPPORTED',
+    ],
+  },
+  local message = { type: 'string' },
+  local detail = { type: 'string' },
```
Comment (stevvooe) [inline schema/errors.libsonnet]: Detail can be anything, not just a string.

File: Makefile [hunk 4]
```diff
@@ -15,3 +15,20 @@ install.tools: .install.gitvalidation
...
+schemas:
+	find schema -path 'schema/test-fixtures' -prune -o -name generate.jsonnet -execdir jsonnet -J jsonnet -m . {} \;
+
+.PHONY: embed-schemas
+embed-schemas:
+	printf "%s\n\n%s\n" "$$(cat .header)" "$$(go generate schema/gen.go)" > schema/fs.go
+
+.PHONY: test-fixtures
+test-fixtures:
+	find schema -path 'schema/test-fixtures/**' -name generate.jsonnet -execdir jsonnet -J jsonnet -m . {} \;
```
Comment (stevvooe) [inline Makefile]: Why use find here?

File: Makefile [hunk 5]
```diff
@@ -15,3 +15,20 @@ install.tools: .install.gitvalidation
 
 .install.gitvalidation:
 	go get -u github.com/vbatts/git-validation
+
+# Generate json-schemas.
+.PHONY: schemas
+schemas:
+	find schema -path 'schema/test-fixtures' -prune -o -name generate.jsonnet -execdir jsonnet -J jsonnet -m . {} \;
```
Comment (stevvooe) [inline Makefile]: Again, just use make input variable for the target. No need to use find.

**Comment by stevvooe** [comment]
Why do we need jsonschema to validate like three fields? This PR adds 1000 lines of code to define the errors and catalog endpoint, and I can't really tell if its right or not. It also strictly defines the error codes and details type, which aren't an exact set.

File: Makefile (hunk 4 above)
Comment (atlaskerr) [reply Makefile]: This was a copypasta from my other project where I had a bunch of `generate.jsonnet` files in the schema directory. I can fix that.

**Comment by atlaskerr** [comment]
@vbatts Yeah keeping everything in lockstep is kind of a drag but I think it's important to have json stuff defined using some kind of standard. Do you think it would be better to have an `oci-schemas` repository that mantains jsonschemas for all OCI specs?

> Why do we need jsonschema to validate like three fields? This PR adds 1000 lines of code to define the errors and catalog endpoint, and I can't really tell if its right or not. It also strictly defines the error codes and details type, which aren't an exact set.
I hear you. It is kind of overkill. I made this PR back when I believed distribution needed to define more custom schemas instead of just exposing the image-spec stuff so I went ahead and setup the foundation.

File: schema/errors.libsonnet (hunk 1 above)
Comment (jzelinskie) [reply schema/errors.libsonnet]: There are multiple ways to handle this in JSONSchema using `anyOf` or `oneOf`.

**Comment by jzelinskie** [comment]
Are we getting a lot of value out of the layer of indirection using jsonnet?
I'm not opposed to having a JSONSchema validation, but I think we need to KISS.

**Comment by atlaskerr** [comment]
> Are we getting a lot of value out of the layer of indirection using jsonnet?

I used jsonnet simply because it's my default tool for working with json. I like having JSONSchema validation but would prefer not to have to work with raw json moving forward.

**Comment by vbatts** [comment]
tbh jsonschema is painful either way. The biggest thing is to not increase the new formats to learn and maintain

**Comment by vbatts** [comment]
atlas, I know this was a load of work. I have wrangled with jsonSchema and openAPI before. Always with big hopes, but consistently found that it became overkill, and unfortunately unless it became a strict gating process that was central to CI and validation then it soon became stale and ignored. 
I'm closing this PR as there is not consensus to make this approach a central validation. I do greatly appreciate your effort in investigating this.

**Pull request closed without merging**